pydantic_ai_slim/pydantic_ai/prompt.py

```diff
@@ -0,0 +1,184 @@
+from __future__ import annotations as _annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Iterable
+from dataclasses import asdict, is_dataclass
+from functools import partial
+from typing import Any, Literal, Union, cast, final
+
+from pydantic import BaseModel
+
+__all__ = (
+    'format_tag',
+    'format_context',
+    'format_examples',
+    'format_rules',
+    'Dialect',
+)
+
+Dialect = Literal['xml']
+BasicType = Union[str, int, float, bool]
+
+
+def format_tag(content: Any, tag: str, dialect: Dialect = 'xml') -> str:
+    """Format content into a tagged string representation using the specified dialect.
+
+    Args:
+        tag: The tag name to wrap the content with.
+        content: The content to be formatted. Can be:
+            - Basic types (str, int, float, bool)
+            - Dictionaries
+            - Lists, tuples, or other iterables
+            - Pydantic models
+            - Dataclasses
+        dialect: The formatting dialect to use. Currently supports:
+            - 'xml': XML format
+
+    Returns:
+        A string representation of the tagged content in the specified dialect.
+
+    Raises:
+        ValueError: If an unsupported dialect is specified.
+        TypeError: If content is of an unsupported type.
+
+    Examples:
+        >>> format_tag({'name': 'John', 'age': 30}, tag='user')
+        '<user><name>John</name><age>30</age></user>'
+
+        >>> from dataclasses import dataclass
+        >>> @dataclass
+        ... class User:
+        ...     name: str
+        ...     age: int
+        >>> user = User('John', 30)
+        >>> format_tag('user', user)
+        '<user><name>John</name><age>30</age></user>'
+    """
+    try:
+        builder = {
+            'xml': XMLTagBuilder,
+        }[dialect](tag, content)
+    except KeyError:
+        raise ValueError(f'Unsupported dialect: {dialect}')
+
+    return builder.build()
+
+
+format_examples = partial(format_tag, tag='examples')
+format_rules = partial(format_tag, tag='rules')
+format_context = partial(format_tag, tag='context')
+
+
+def _normalize_type(content: Any) -> dict[str, Any] | list[Any] | BasicType:
+    if isinstance(content, BaseModel):
+        return content.model_dump()
+    elif is_dataclass(content):
+        return asdict(cast(Any, content))
+    elif isinstance(content, Iterable) and not isinstance(content, (str, dict)):
+        return list(cast(Iterable[Any], content))
+    elif isinstance(content, (str, int, float, bool, dict)):
+        return cast(Union[BasicType, dict[str, Any]], content)
+    else:
+        raise TypeError(f'Unsupported content type: {type(content)}')
+
+
+class TagBuilder(ABC):
```

Comment:
I don't think we need all this, better to just have a good implementation of xml for now.